source/extensions/filters/network/postgres_proxy/postgres_message.h

```diff
@@ -0,0 +1,319 @@
+namespace PostgresProxy {
+
+/**
+ * Postgres messages are described in official Postgres documentation:
+ * https://www.postgresql.org/docs/12/protocol-message-formats.html
+ *
+ * Most of messages start with 1-byte message identifier followed by 4-bytes length field. Few
+ * messages are defined without starting 1-byte character and are used during well-defined initial
+ * stage of connection process.
+ *
+ * Messages are composed from various fields: 8, 16, 32-bit integers, String, Arrays, etc.
+ *
+ * Structures defined below have the same naming as types used in official Postgres documentation.
+ *
+ * Each structure has the following methods:
+ * read - to read number of bytes from received buffer. The number of bytes depends on structure
+ * type. toString - method returns displayable representation of the structure value.
+ *
+ */
+
+// Template for integer types.
+// Size of integer types is fixed and depends on the type of integer.
+template <typename T> class Int {
+public:
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
+    if ((data.length() - pos) < sizeof(T)) {
+      return false;
+    }
+    value_ = data.peekBEInt<T>(pos);
+    pos += sizeof(T);
+    left -= sizeof(T);
+    return true;
+  }
+
+  std::string toString() const { return fmt::format(getFormat(), value_); }
+
+  constexpr const char* getFormat() const { return "[{}]"; }
+  T get() const { return value_; }
+
+private:
+  T value_{};
+};
+
+using Int32 = Int<uint32_t>;
+using Int16 = Int<uint16_t>;
+using Int8 = Int<uint8_t>;
+
+// 8-bits character value.
+using Byte1 = Int<char>;
+
+// String type requires byte with zero value to indicate end of string.
+class String {
+public:
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
+    // First find the terminating zero.
+    char zero = 0;
+    auto index = data.search(&zero, 1, pos);
+    if (index == -1) {
+      return false;
+    }
+
+    // Reserve that many bytes in the string.
+    auto size = index - pos;
+    value_.resize(size);
+    // Now copy from buffer to string.
+    data.copyOut(pos, index - pos, value_.data());
+    pos += (size + 1);
+    left -= (size + 1);
+
+    return true;
+  }
+
+  std::string toString() const { return absl::StrCat("[", value_, "]"); }
+
+private:
+  std::string value_;
+};
+
+// ByteN type is used as the last type in the Postgres message and contains
+// sequence of bytes. The length must be deduced from message length.
+class ByteN {
+public:
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
+    if (left > (data.length() - pos)) {
+      return false;
+    }
+    value_.resize(left);
+    data.copyOut(pos, left, value_.data());
+    pos += left;
+    left = 0;
+    return true;
+  }
+
+  std::string toString() const {
+    std::string out = "[";
+    bool first = true;
+    for (const auto& i : value_) {
+      std::string buf;
+      if (first) {
+        buf = fmt::format("{}", i);
```

Comment:
no need to call format() here; you can just do absl::StrAppend(&out, i)

Reply:
Thanks for pointing it out. I thought that absl::StrAppend takes only strings as arguments.